src/dsl/variables.rs

```diff
@@ -134,56 +135,1135 @@ macro_rules! implement_boundable {
+        LpExpression {
+            root: root,
+            arena: arena
+        }
+    }
+
+    pub(crate) fn get_root_index(&self) -> LpExprArenaIndex {
+        self.root
+    }
+
+    fn set_root_to_index(&mut self, root_index: LpExprArenaIndex) {
+        self.root = root_index;
+    }
+
+    fn push_as_expr<T>(&mut self, lp_expr: &T) -> LpExprArenaIndex where T: Into<LpExprNode> + Clone {
+        let index = self.arena.len();
+        self.arena.push(lp_expr.clone().into());
+        return index
+    }
+
+    fn clone_expr_at_and_push(&mut self, index: LpExprArenaIndex) -> LpExprArenaIndex {
+        let new_index = self.arena.len();
+        self.arena.push(self.expr_clone_at(index));
+        return new_index
+    }
+
+    fn overwrite_expr_at(&mut self, index: LpExprArenaIndex, lp_expr: LpExprNode) {
+       self.arena[index] = lp_expr;
+    }
+
+    pub(crate) fn expr_ref_at(&self, index: LpExprArenaIndex) -> &LpExprNode {
+        match self.arena.get(index) {
+            Some(expr) => expr,
+            None => panic!("Requested index out of bound of LpExpression vector. This should not happen.")
+        }
+    }
+
+    fn expr_clone_at(&self, index: LpExprArenaIndex) -> LpExprNode {
+        match self.arena.get(index) {
+            Some(expr) => expr.clone(),
+            None => panic!("Requested index out of bound of LpExpression vector. This should not happen.")
+        }
+    }
+
+    pub(crate) fn get_root_expr_ref(&self) -> &LpExprNode {
+        self.expr_ref_at(self.root)
+    }
+
+    pub(crate) fn split_off_constant(&mut self) -> f32 {
+        match self.expr_clone_at(self.root) {
+            LitVal(c) => {
+                self.clone_from(&LpExpression::new());
+                c
+            },
+            LpCompExpr(Addition, e1, e2) => {
+                if let LitVal(c) = self.expr_clone_at(e2) {
+                    self.set_root_to_index(e1);
+                    c
+                } else {
+                    0.0
+                }
+            },
+            LpCompExpr(Subtraction, e1, e2) => {
+                if let LitVal(c) = self.expr_clone_at(e2) {
+                    self.set_root_to_index(e1);
+                    -c
+                } else {
+                    0.0
+                }
+            },
+            _ => 0.0
+        }
+    }
+
+    /// Recursively unwrap an arena of expressions.
+    /// Assumes that self.simplify() was called on the arena before starting recursion, here. That way,
+    /// `LitVal()`s of nested Multiplication expressions are already accumulated towards the left side
+    /// and `LitVal()`s of all Addition / Subtraction expressions are already accumulated towards the
+    /// right end of the tree, with all Addition / SUbtraction expressions on the most global level
+    /// possible.
+    #[cfg(feature = "native_coin_cbc")]
+    pub(crate) fn var_lit(&self, expr_index: LpExprArenaIndex, lst: &mut Vec<(String, f32)>, mul: f32) {
```

Comment:
Now that `var_lit` is more public than before, I'd go for some changes:
1. The name makes kind of sense to _me_, but it may not sound very semantic to others. I don't know what you think, feel free to change it.
2. I was passing `lst` as an argument out of lazyness. I believe that a better interface would be to just return a `Vec` and remove it from the arguments.
3. What do you think of calling `self.simplify()` inside of `var_lit`? Would it introduce too much overhead for the cases when the Expression was already simplified? I guess that would require to pass `&mut self`, I am not sure about this 3rd point.

Reply:
Thanks for the detailed feedback!

I just realized, that `var_lit()` does not need to be public at all: I simply moved it back into `native_cbc.rs` and extend the implementation of `LpExpression` over there (see [the newest commit](https://github.com/jcavat/rust-lp-modeler/pull/61/commits/178eec7077a4373c81a898af7c08675e4ddd3bdf)). That way, no separate feature gate declaration is needed and the function does not even need to be `pub(crate)`, but can stay private to that file/module.

Regarding the naming, the function basically creates a vector of variables where each entry is the name of the variable and its associated factor, so all the ingredients you need to write down the sum that the expression represents, right? So maybe something like `expr_to_var_name_vec_with_factors()`? A bit long, but very descriptive and in the few cases someone will have to type it out, they will hopefully be assisted by some IDE. Or do you have suggestions for further improvement?

Regarding the `lst` being passed in: This does make sense, as the function is recursive. Thus, we need a data structure that exists beyond the execution of each individual call. The alternative would be to turn the function from recursive to something for-loop based, with a stack of expressions to still handle and a stack for accumulating the signs of factors. This could also become necessary for very long expressions, where recursion might overflow the stack as we had for `LpExpression::simplify`, the reason for PR #48 .

And regarding calling `self.simplify()` inside of `var_lit()`, this is kind of analogous to the previous point. As long as the function is recursive, this would mean that `self.simplify()` gets called on every recursion, creating considerable (unnecessary) overhead. Also, it would require `var_lit()` to work on `&mut self` instead of `&self` and would have the side-effect of altering self, something that users might not expect (although `var_lit()` is now back to being restricted to `native_cbc.rs`...). But turning it into a non-recursive for-loop function might change the coordinates for this point, as well.

Reply:
OK, now that it is back to *native_cbc.rs*, I am fine as how it is. That's how I would do it if it was public for the crate:

1. `expr_to_var_name_vec_with_factors` sounds fine to me!
2. My idea to return a `Vec<_>` instead of mutate it was about wrapping it inside another function. Something like:
```rust
fn expr_to_var_name_vec_with_factors(&self, expr_index: LpExprArenaIndex) -> Vec<(String, f32)> {
    // inner function
    fn _recursive_to_var_with_factors(expr_index: LpExprArenaIndex, lst: Vec<(String, f32)>, mul: f32) {
        // same as previous var_list
    }
    // simplify here
    // ...
    // call inner funtion
    let lst: Vec<(String, f32)> = Vec::new();
    _recursive_to_var_with_factors(expr_index, lst, 1.);
    lst
}
```
3. The same as 2 (call inside the wrapper and then call the inner recursive one.

If you feel like doing so, change **1** but I wouldn't bother with **2** and **3**.